File: hw3-1.jl
function findSum(inp::AbstractVector, sum::Int) :: Union{Vector{Tuple{Int,Int}}, Nothing}
    if len(inp) < 2 || typeof(inp) != Vector{Int} return nothing end
    
    sets = pair(inp); ret = Tuple{Int,Int}[]
    for i in sets
        if i[1]+i[2] == sum
            push!(ret, i)
        end
    end
    trim(ret)
end

function extendedFindSum(inp::AbstractVector) :: Union{Vector{Tuple{Int, Vector{Tuple{Int,Int}}}}, Nothing}
    if len(inp) < 2 || typeof(inp) != Vector{Int} return nothing end

    sets = pair(inp); ret = Tuple{Int, Tuple{Int,Int}}[]
    for i in sets
        push!(ret, (i[1]+i[2],i))
    end
    sort!(classify(trim(ret)))
end

function len(inp::AbstractVector) :: Int
    ret = 0
    for i in inp
        ret += 1
    end
    ret
end

function pair(inp::Vector{Int}) :: Vector{Tuple{Int,Int}}
    sets = Any[]; push!(sets, Int[])
    for i in inp, j in 1:len(sets)
        push!(sets, [sets[j]; i])
    end

    ret = Tuple{Int,Int}[]
    for i in sets
        if len(i) == 2
            push!(ret, (i[1],i[2]))
        end
    end
    ret
end

function trim(inp::AbstractVector) :: AbstractVector
    if len(inp) == 0 return [] end
    
    head = popfirst!(inp); tail = trim(inp)
    if typeof(inp) == Vector{Tuple{Int,Int}}
        ret = Tuple{Int,Int}[]
    elseif typeof(inp) == Vector{Tuple{Int,Tuple{Int,Int}}}
        ret = Tuple{Int,Tuple{Int,Int}}[]
    end

    push!(ret, head)
    for i in tail
        if i != head
            push!(ret, i)
        end
    end
    ret
end

function classify(inp::Vector{Tuple{Int,Tuple{Int,Int}}}) :: Vector{Tuple{Int, Vector{Tuple{Int,Int}}}}
    dict = Dict{Int, Vector{Tuple{Int,Int}}}()
    map(inp) do tup
        if existkey(dict, tup[1])
            push!(dict[tup[1]], tup[2])
        else
            xs = Tuple{Int,Int}[]
            dict[tup[1]] = push!(xs, tup[2])
        end
    end
    
    ret = Tuple{Int, Vector{Tuple{Int,Int}}}[]
    for i in dict
        push!(ret, (i[1],i[2]))
    end
    ret
end

function existkey(inp::AbstractDict, key::Any) :: Bool
    for (k, v) in inp
        if key == k return true end
    end
    false
end

function sort!(inp::Vector{Tuple{Int,Vector{Tuple{Int,Int}}}}, start::Int = 1, stop::Int = len(inp)) :: Vector{Tuple{Int,Vector{Tuple{Int,Int}}}}
    if stop > start
        left, right, pivot = start, stop, inp[rand(start:stop)][1]
        while left <= right
            while inp[left][1] < pivot
                left += 1
            end
            while inp[right][1] > pivot
                right -= 1
            end
            if left <= right
                inp[left], inp[right] = inp[right], inp[left]
                left += 1; right -= 1
            end
        end
        sort!(inp, start, right)
        sort!(inp, left, stop)
    end
    inp
end

function main() :: Nothing
    println("findSum:")
    show(findSum([1,3,8,12,7,11,9,4,2,10,5], 12))
    println("\nextendedFindSum:")
    show(extendedFindSum([1,3,8,12,7,11,9,4,2,10,5]))
    println()
end

main()

File: Haskell/hw3_1.hs
findSum :: [Int] -> Int -> [(Int, Int)]
findSum xs sum = trim [(x, y) | ys <- sets xs
                              , len ys == 2
                              , let x = head ys
                              , let y = head (tail ys)
                              , (+ x) y == sum]

extendedFindSum :: [Int] -> [(Int, [(Int, Int)])]
extendedFindSum xs = merge $ trim [(sum, (x, y)) | ys <- sets xs
                                                 , len ys == 2
                                                 , let x = head ys
                                                 , let y = head (tail ys)
                                                 , let sum = (+ x) y]

len :: [a] -> Int
len = foldr (\_ n -> 1+n) 0

trim :: Eq a => [a] -> [a]
trim [] = []
trim (x:xs) = x : [y | y <- trim xs, x /= y]

merge :: [(Int, (Int, Int))] -> [(Int, [(Int, Int)])]
merge xs = do let parse = \ls -> (fst $ head ls, map snd ls)
              map parse $ classify (\a b -> fst a == fst b) $ sort xs

sort :: (Ord a) => [(a, (a, a))] -> [(a, (a, a))]
sort [] = []
sort (x:xs) = sort [y | y <- xs, fst y <= fst x]
              ++ [x] ++
              sort [y | y <- xs, fst y > fst x]
                    
classify :: (a -> a -> Bool) -> [a] -> [[a]]
classify _ []       = []
classify cmp (x:xs) = let (ts,fs) = span (cmp x) xs
                      in (x:ts) : classify cmp fs

sets :: [a] -> [[a]]
sets xs = [] : valSets xs

valSets :: [a] -> [[a]]
valSets sets = case sets of
                 [] -> []
                 (x:xs) -> let acc `op` ret = acc : (x:acc) : ret
                           in [x] : foldr op [] (valSets xs)

main :: IO ()
main = do putStrLn "findSum:"
          print(findSum [1,3,8,12,7,11,9,4,2,10,5] 12)
          putStrLn "extendedFindSum:"
          print(extendedFindSum [1,3,8,12,7,11,9,4,2,10,5])